lib/sleeper.ts

```diff
@@ -0,0 +1,138 @@
+import assert from "assert";
+
+import log from "./log";
+
+/**
+ * What state is our `Sleeper` in?
+ *
+ * This type union uses TypeScript magic to work like a Rust enumeration. When
+ * we compare `this.state.name` against a known state name, TypeScript narrows
+ * down the fields available on our state.
+ *
+ * We use this relatively heavyweight state tracking representation because we
+ * have a number of member variables that are only present in certain states,
+ * and we want to be very careful about how we manage them.
+ */
+type SleeperState = ReadyState | SleepingState | CanceledForAllTimeState;
+
+/** We are ready to sleep. */
+type ReadyState = {
+  /** The name of this state. */
+  name: "READY";
+};
+
+/** We are currently sleeping. */
+type SleepingState = {
+  /** The name of this state. */
+  name: "SLEEPING";
+
+  /** The timeout that will fire when the sleep ends. */
+  timeoutHandle: NodeJS.Timeout;
+
+  /**
+   * Somebody is calling `await sleeper.sleepSecs`. Calling this will wake them
+   * up.
+   */
+  resolve: () => void;
+};
+
+/** We will never sleep again. */
+type CanceledForAllTimeState = {
+  /** The name of this state. */
+  name: "CANCELED_FOR_ALL_TIME";
+};
+
+/** A utility class that provides interruptible sleep. */
+export class Sleeper {
+  /** The state of this `Sleeper`. */
+  private state: SleeperState;
+
+  /**
+   * Create a new `Sleeper`.
+   *
+   * ```
+   * const sleeper = new Sleeper();
+   * await sleeper.sleepSecs(30);
+   *
+   * // In another task.
+   * sleeper.cancelCurrentAndFutureSleeps();
+   * ```
+   */
+  constructor() {
+    this.state = { name: "READY" };
+  }
+
+  /**
+   * Sleep for the specified number of seconds. This may be interrupted using
+   * `cancelCurrentAndFutureSleeps`. Only one sleep may be running at a time.
+   *
+   * @param secs The number of seconds to sleep.
+   */
+  sleepSecs(secs: number): Promise<void> {
+    log.trace("sleepSecs", secs, "in", this.state.name);
+
+    // If we've been canceled, don't even start sleeping.
+    if (this.state.name === "CANCELED_FOR_ALL_TIME") {
+      // Return a pre-resolved promise.
+      return Promise.resolve();
+    }
+
+    // Install a timeout with a callback.
+    assert(
+      this.state.name === "READY",
+      `cannot sleep in state ${JSON.stringify(this.state)}`
+    );
+    return new Promise((resolve) => {
+      this.state = {
+        name: "SLEEPING",
+        resolve,
+        timeoutHandle: setTimeout(() => {
+          this.wakeUpAfterTimeout();
+        }, secs * 1000),
+      };
+    });
+  }
```

Comment:
wow this got complicated

Reply:
Yeah, cleaning up timeouts turns out to be important—if we don't, then process shutdown hangs until all the timeouts fire.

I like this version of the code because it's easy for me to see that all the state transitions are correct by inspection.